docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,189 @@
+---
+title: agent-upgrade
+authors:
+  - "@jhernand"
+creation-date: 2022-06-13
+last-updated: 2022-06-13
+---
+
+# Agent upgrade
+
+## Summary
+
+Usually the time since the agent is started and the cluster is installed is
+short. But in some situations it may be long, even many days, and during that
+long time the service may be upgraded and require a newer version of the agent.
+The current mechanism to perform that agent upgrade is to reboot the host. That
+is inconvenient, specially if there are many hosts or if rebooting a host
+requires manual intervention. This is a proposal to implement a mechanism to
+upgrade the agent without having to reboot the host.
+
+## Motivation
+
+Reduce the need to reboot hosts when there are changes in the service that
+require new versions of the agent.
+
+### Goals
+
+A host that has been booted and registered with matching versions of the agent
+and the service will not need to be rebooted in order to install the cluster
+with a newer version of the service.
+
+Note that in this context _version_ is really a container image name, like
+`registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141`.
+
+### Non-Goals
+
+Hosts will still need to be rebooted with a new version of the ISO if there
+are changes in the part of the agent that performs the initial registration.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to be able to generate an ISO, boot and register
+hosts with a version of the service, then upgrade the service to a version that
+requires a newer version of the agent, and finally install the cluster without
+having to reboot the hosts.
+
+#### Story 2
+
+As a cluster Creator I want to be able to generate an ISO, boot and register
+hosts with a version of the service, then downgrade the service to a version
+that requires an older version of the agent, and finally install the cluster
+without having to reboot the hosts.
```

Comment:
I guess the main point here is that I'd like to be able to successfully install the cluster even if cases the agent image has to be changed due to the service requirements, whether if it is an "upgrade" change, or a "downgrade" one.

Reply:
I removed these two stories and replaced them one a single one explaining that the developer expects the agent to be upgraded automatically so that she can do breaking changes to both the service and the agent. Hope that is clearer.